### PR TITLE
Add support for basic authorization

### PR DESCRIPTION
**Pull request by andrewheiss**
With self-hosted ntfy installations (and maybe with ntfy.sh?), you can create password-protected topics. To send notifications to these topics, you have to pass a basic authorization header in the POST/PUT/whatever request.

This PR adds two new environment variables: `NTFY_USERNAME` and `NTFY_PASSWORD`. If they're set, it uses a new function named `ntfy_authorization()` to generate the correct base64-encoded HTTP header, which is then passed to `httr::POST()` from inside `ntfy::send()`. If the environment variables are empty, everything works like normal.

For example, given a username of "example" and a password of "super-secret-password", `ntfy_authorization()` will create the following httr request object:

```r
Sys.setenv("NTFY_USERNAME" = "example")
Sys.setenv("NTFY_PASSWORD" = "super-secret-password")

# This generates the correct heading:
ntfy_authorization()
#> <request>
#> Headers:
#>   * Authorization: Basic ZXhhbXBsZTpzdXBlci1zZWNyZXQtcGFzc3dvcmQ
```

All someone needs to do to use a protected topic is set the correct environment variables:

```r
# Or set these in .Renviron
Sys.setenv("NTFY_TOPIC" = "protected-topic")
Sys.setenv("NTFY_SERVER" = "https://ntfy.example.com")
Sys.setenv("NTFY_USERNAME" = "example")
Sys.setenv("NTFY_PASSWORD" = "super-secret-password")

library(ntfy)
ntfy_send(message = "Posting to a protected topic")
```

### CODE REVIEW

**Comment by jonocarroll** [comment]
I'll add a couple of inline comments as a review, but this looks great - thanks!

Now I start to wish I'd used {httr2} because that has `req_auth_basic()`...

```r
library(httr2)
request("http://ntfy.sh") |> 
   httr2::req_auth_basic("example", "super-secret-password") |> 
   print(redact_headers = FALSE)
# <httr2_request>
# GET http://ntfy.sh
# Headers:
# • Authorization: 'Basic ZXhhbXBsZTpzdXBlci1zZWNyZXQtcGFzc3dvcmQ='
# Body: empty
```

but there's nothing wrong with spelling it out with `add_headers()`.

File: R/ntfy.R [hunk 1]
```diff
@@ -69,9 +104,9 @@ ntfy_send <- function(message  = "test",
   payload <- Filter(Negate(is.null), payload)
 
   if (!is.null(image)) {
-    send_image(trim_slash(server), topic, image, payload, ...)
+    send_image(trim_slash(server), topic, image, payload, config = auth, ...)
```
Comment (jonocarroll) [inline R/ntfy.R:107]: Does `send_image` need an explicit `config` argument? That function already constructs a `httr_header`... does this interfere when sending an image? I'll also test.

File: R/ntfy.R [hunk 2]
```diff
@@ -18,6 +18,39 @@ ntfy_server <- function(var = "NTFY_SERVER") {
...
+#' #> Headers:
+#' #>   * Authorization: Basic ZXhhbXBsZTpzdXBlci1zZWNyZXQtcGFzc3dvcmQ
+#' }
+#' 
+ntfy_authorization <- function(var_username = "NTFY_USERNAME", var_password = "NTFY_PASSWORD") {
+  username = Sys.getenv(var_username)
+  password = Sys.getenv(var_password)
+  
+  if (username == "") {
+    httr::add_headers()
```
Comment (jonocarroll) [inline R/ntfy.R:46]: What about adding an alias function 
```r
no_auth = httr::add_headers
```
that can alternatively be passed to `ntfy_send()` in place of `ntfy_authorization(var_username = "")` in order to disable auth (even though the env var is set)?

Another option would be to allow `ntfy_send(auth=FALSE)` and handle the logic there.

File: R/ntfy.R [hunk 3]
```diff
@@ -18,6 +18,39 @@ ntfy_server <- function(var = "NTFY_SERVER") {
...
+#' 
+#' @return an object with class `request` that is passed to [httr::POST()]
+#' @export
+#' 
+#' @examples
+#' \dontrun{
+#' Sys.setenv("NTFY_USERNAME" = "example")
+#' Sys.setenv("NTFY_PASSWORD" = "super-secret-password")
+#' 
+#' # This generates the correct heading:
```
Comment (jonocarroll) [inline R/ntfy.R]:
```suggestion
#' # This generates the correct header:
```

**Review by jonocarroll — CHANGES_REQUESTED** [review]
Looks great, but I'm unable to test any of the auth pieces on the public server (returns 401: unauthorized). If you know of a self-hosted server with auth I could try out then please let me know, otherwise I'll need to find some time to spin one up.

The env vars are a handy way to make the workflow clean for users, but I'd like an escape hatch should they not want to use them (currently explicitly passing `topic` or `server` overrides) - turning _off_ auth requires passing through `ntfy_authorization()` at the moment and that could perhaps be simplified for that use-case.

Many thanks for implementing this!

**Comment by andrewheiss** [comment]
I have a custom server at https://ntfy.andrewheiss.com and there's a user named `example` with the password `super-secret-password` that has access to a channel named `r-testing`, so this should work for testing:

```r
Sys.setenv("NTFY_TOPIC" = "r-testing")
Sys.setenv("NTFY_SERVER" = "https://ntfy.andrewheiss.com")
Sys.setenv("NTFY_USERNAME" = "example")
Sys.setenv("NTFY_PASSWORD" = "super-secret-password")

library(ntfy)
ntfy_send(title = "Testing", message = "This should work", tags = "crossed_fingers")
```

**Comment by jonocarroll** [comment]
On testing: 

* sending both the plain message and image to the authorized server works great, cheers!
* switching back to the default server with `Sys.setenv(NTFY_SERVER="")` and `Sys.setenv(NFTY_TOPIC="mytopic")` I get a failure, initially because it still needs an actual server, and {ntfy} sets that to the default in `onLoad` if one isn't specified - I'll add a TODO to add a `reset_server()` so that users don't need to know exactly which URL is required.
* With that fixed, it still fails with 401:unauthorized because `NTFY_USERNAME` is still set so it's passing the auth header.
* Turning that off with `ntfy_send("username set but no auth needed", auth = httr::add_headers())` it once again works. Likewise with an image. I think supporting `auth=FALSE` or `auth=no_auth()` (where the latter is more type-safe) would be useful here. Given that the public server does _not_ have auth, we could also safely disable auth when the server is the public one.

File: R/ntfy.R (hunk 1 above)
Comment (jonocarroll) [reply R/ntfy.R:107]: It seems this "just works" via the `...` magic.

Comment (andrewheiss) [reply R/ntfy.R:107]: Yeah, this surprised me when I first wrote it. In the documentation for the different HTTP verbs in {httr} it says 

```
...    Further named parameters, such as query, path, etc, passed on to modify_url(). Unnamed 
       parameters will be combined with config().
```

So I think everything extra that gets passed to POST/GET/PUT/etc. somehow gets lumped into a `config` list, including preexisting config arguments.

File: R/ntfy.R (hunk 3 above)
Comment (andrewheiss) [reply R/ntfy.R]: Fixed!

**Comment by andrewheiss** [comment]
I'm guessing that the majority of use cases use the public auth-less server, so it's probably best to make it work without authorization by default. I like the httr2 approach of having the request piped into an authorization function—I might actually try to update everything to httr2 instead (there are only a handful of GET/POST/PUT functions in the package), because then something like this would work without needing to unset environment variables or anything:

```r
library(httr2)

# If authorization is required, add the header to the request, otherwise just
# pass the request object through
req_add_auth <- function(req, auth) {
  if (auth) {
    httr2::req_auth_basic(req, "example", "super-secret-password")
  } else {
    req
  }
}

auth <- FALSE  # Set this in the ntfy_send() function or wherever else
resp <- request("http://ntfy.sh") |>
  req_method("POST") |> 
  req_add_auth(auth) |> 
  print(redact_headers = FALSE)
#> <httr2_request>
#> POST http://ntfy.sh
#> Body: empty

auth <- TRUE
resp <- request("http://ntfy.sh") |>
  req_method("POST") |> 
  req_add_auth(auth) |> 
  print(redact_headers = FALSE)
#> <httr2_request>
#> POST http://ntfy.sh
#> Headers:
#> • Authorization: 'Basic ZXhhbXBsZTpzdXBlci1zZWNyZXQtcGFzc3dvcmQ='
#> Body: empty
```

**Comment by jonocarroll** [comment]
I'm fine with a migration to `httr2` being a distinct thing, so using

```r
ntfy_send([args], auth = NULL)
```

as the default with the possibility of passing in this implementation of `auth = ntfy_authorization()` would be fine, and could be refactored afterwards.

**Comment by jonocarroll** [comment]
One more task to add - please add a `ctb` entry to `Authors@R` in `DESCRIPTION`.

File: R/ntfy.R (hunk 2 above)
Comment (andrewheiss) [reply R/ntfy.R:46]: Here I made `auth = NULL` by default, and no authorization attempt is made unless the user specifies `auth = ntfy_authorization()`: https://github.com/jonocarroll/ntfy/pull/12/commits/44d6c8db362db66ab0b6b3dd2f5a6bde41f95777

**Comment by jonocarroll** [comment]
I added the same `auth` argument to the other functions so that e.g. `ntfy_history(auth = ntfy_authorization())` works (and it _does_ return 200, it just doesn't get processed correctly - a different bug to fix; something wrong with the `fromJSON` bit, even from the public server).

**Pull request merged**